File: problems/Longest Palindromic Subsequence/longestPalindromeSubseq.cs
public class Solution {
    public int LongestPalindromeSubseq(string s) {
        int[,] dp = new int[s.Length, s.Length];

        return lpsTopDown(s, dp, 0, s.Length - 1);
    }

    private int lpsTopDown(string s, int[,] dp, int start, int end) {
        if (start > end) {
            return 0;
        }

        if (start == end) {
            return 1;
        }

        if (0 == dp[start, end]) {
            if (s[start] == s[end]) {
                dp[start, end] = 2 + lpsTopDown(s, dp, 1 + start, end - 1);
            } else {
                int lpsLeft = lpsTopDown(s, dp, 1 + start, end);
                int lpsRight = lpsTopDown(s, dp, start, end - 1);

                dp[start, end] = Math.Max(lpsLeft, lpsRight);
            }
        }

        return dp[start, end];
    }
}

File: problems/Longest Palindromic Subsequence/longestPalindromeSubseq.java
class Solution {
    public int longestPalindromeSubseq(String s) {
        int n = s.length();
        int[][] dp = new int[n][n];

        for (int start = 0; n > start; ++start) {
            for (int end = start; 0 <= end; --end) {
                if (start == end) {
                    dp[start][end] = 1;
                } else if (s.charAt(start) == s.charAt(end)) {
                    dp[start][end] = 2 + dp[start - 1][1 + end];
                } else {
                    dp[start][end] = Math.max(dp[start - 1][end], dp[start][1 + end]);
                }
            }
        }

        return dp[n - 1][0];
    }
}
